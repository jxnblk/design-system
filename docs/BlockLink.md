# BlockLink

`<BlockLink />` can be used as an anchor link element (`<a>`) as a wrapping parent to compose its block-level children components (supported by HTML 5). `<BlockLink />` is a styled-components wrapper of `<Link />` component to remove text-decoration and colors styles and sets display as block.

```.jsx
<BlockLink href="https://www.priceline.com" target="_blank">
  <BackgroundImage image='https://source.unsplash.com/bITjK6W2Alw/1024x768?q=20' width="640px">
    <Box p={4}>
      <Text align="center" color="white">Click to open priceline.com in new tab!</Text>
    </Box>
  </BackgroundImage>
</BlockLink>
```
