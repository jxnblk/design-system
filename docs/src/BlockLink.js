import styled from 'styled-components'

const hoverColor = props =>
  props.hoverColor
    ? {
        '&:hover': {
          color: props.theme.colors[props.hoverColor] || props.hoverColor
        }
      }
    : null

const BlockLink = styled.a`
  display: block;
  color: inherit;
  text-decoration: none;
  ${hoverColor};
`

export default BlockLink
