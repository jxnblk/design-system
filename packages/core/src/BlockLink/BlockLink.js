import styled from 'styled-components'

import { Link } from '../Link'
import { mapProps, refPropType, getSCMigrationRef } from '../utils'

const BlockLink = mapProps(({ dsRef, ...props }) => ({
  [getSCMigrationRef()]: dsRef,
  ...props
}))(styled(Link)`
  display: block;
  color: inherit;
  text-decoration: none;
`)

BlockLink.displayName = 'BlockLink'

BlockLink.propTypes = {
  dsRef: refPropType
}

export default BlockLink
