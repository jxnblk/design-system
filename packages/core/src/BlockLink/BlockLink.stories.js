import React from 'react'
import { storiesOf } from '@storybook/react'

import { BackgroundImage, BlockLink, Box, Flex, Text } from '..'
import ForwardRefDemo from '../../storybook/utils/ForwardRefsDemo'

const image =
  'https://images.unsplash.com/photo-1446776811953-b23d57bd21aa?ixlib=rb-0.3.5&q=80&fm=jpg&crop=entropy&cs=tinysrgb&w=1080&fit=max&s=aee8a50c86478d935556d865624506e4'

storiesOf('BlockLink', module)
  .add('containing BackgroundImage', () => (
    <Flex justifyContent='center' alignItems='center' color='white'>
      <BlockLink href='https://www.priceline.com' target='_blank'>
        <BackgroundImage image={image} width='640px'>
          <Box p={4}>
            <Text textAlign='center'>
              Click to open priceline.com in new tab!
            </Text>
          </Box>
        </BackgroundImage>
      </BlockLink>
    </Flex>
  ))
  .add('Forward refs', () => (
    <ForwardRefDemo
      refChild={(dsRef) => (
        <>
          <BlockLink color='darkGray' dsRef={dsRef}>
            I'm a link!
          </BlockLink>
          <br />
          <Button
            color='error'
            onClick={() => (dsRef.current.innerHTML = 'Bacon!')}
            mt={4}
          >
            Click to update link text via ref
          </Button>
        </>
      )}
    />
  ))
  .add('composition without container', () => (
    <Flex justifyContent='center' alignItems='center' color='purple'>
      <BlockLink href='https://www.google.com'>
        <Text fontSize={2} bold textAlign='center'>
          Click to go to google.com!
        </Text>
      </BlockLink>
    </Flex>
  ))
