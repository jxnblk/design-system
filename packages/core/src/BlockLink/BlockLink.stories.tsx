import React from 'react'

import { BackgroundImage, BlockLink, Box, Button, Flex, Text } from '..'
import ForwardRefDemo from '../storybook/utils/ForwardRefsDemo'

const image =
  'https://images.unsplash.com/photo-1446776811953-b23d57bd21aa?ixlib=rb-0.3.5&q=80&fm=jpg&crop=entropy&cs=tinysrgb&w=1080&fit=max&s=aee8a50c86478d935556d865624506e4'

export default {
  title: 'BlockLink',
  component: BlockLink,
}

export const ContainingBackgroundImage = () => (
  <Flex justifyContent='center' alignItems='center' color='white'>
    <BlockLink href='https://www.priceline.com' target='_blank'>
      <BackgroundImage image={image} width='640px'>
        <Box p={4}>
          <Text textAlign='center'>Click to open priceline.com in new tab!</Text>
        </Box>
      </BackgroundImage>
    </BlockLink>
  </Flex>
)

ContainingBackgroundImage.story = {
  name: 'containing BackgroundImage',
}

export const ForwardRefs = () => {
  function refChild(dsRef) {
    function onClick() {
      return (dsRef.current.innerHTML = 'Bacon!')
    }
    return (
      <>
        <BlockLink color='text.dark' ref={dsRef}>
          I am a link!
        </BlockLink>
        <br />
        <Button color='error' onClick={onClick} mt={4}>
          Click to update link text via ref
        </Button>
      </>
    )
  }
  return <ForwardRefDemo refChild={refChild} />
}

ForwardRefs.story = {
  name: 'Forward refs',
}

export const CompositionWithoutContainer = () => (
  <Flex justifyContent='center' alignItems='center' color='purple'>
    <BlockLink href='https://www.google.com'>
      <Text fontSize={2} bold textAlign='center'>
        Click to go to google.com!
      </Text>
    </BlockLink>
  </Flex>
)

CompositionWithoutContainer.story = {
  name: 'composition without container',
}
