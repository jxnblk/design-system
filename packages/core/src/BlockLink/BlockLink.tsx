import styled from 'styled-components'
import { Link } from '../Link/Link'

export const BlockLink = styled(Link)`
  display: block;
  color: inherit;
  text-decoration: none;
`

BlockLink.displayName = 'BlockLink'
