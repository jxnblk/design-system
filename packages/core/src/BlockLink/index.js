export { default as BlockLink } from './BlockLink'
