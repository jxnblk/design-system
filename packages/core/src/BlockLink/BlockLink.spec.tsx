import React from 'react'
import { BlockLink } from './BlockLink'

describe('BlockLink', () => {
  test('renders', () => {
    const json = rendererCreateWithTheme(<BlockLink>raw text</BlockLink>).toJSON()
    expect(json).toMatchSnapshot()
  })
})
