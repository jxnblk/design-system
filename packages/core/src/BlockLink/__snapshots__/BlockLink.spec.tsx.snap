// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlockLink renders 1`] = `
.c0 {
  font-family: 'Montserrat','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.4;
  font-weight: 500;
}

.c0 * {
  box-sizing: border-box;
}

.c1 {
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #0068ef;
}

.c1:hover {
  color: #049;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c2 {
  display: block;
  color: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div
  className="c0"
>
  <a
    className="c1 c2"
    color="primary"
    rel={null}
    size="medium"
  >
    raw text
  </a>
</div>
`;
